add support for aliases

binary1230 opened the pull request:
Hey @lins05, thanks for the work on the awesome bot.  We're (Magfest) attempting to use it for devops purposes and shenanigans to replace our existing bot, which is Hubot-based, with a python bot (because we're a python shop mostly).  This will be the first pull request of a few that we're currently working on which is designed to mimic some features that Hubot has which our chat has become used to.  I'm going to try and clean these changes up with the hope of pulling them into your upstream repo.

---

This PR:
You can use shorthand notation or aliases for talking to the bot now.  This allows `slackbot.settings.ALIASES` to take a comma-delimited list of aliases the bot will respond to.

For example, normally you'd have to do:

```
@botuser hello
```

but if you add this env var / add to settings.py:

```
export SLACKBOT_ALIASES='!,$'
./run.py
```

then you can do:

```
! hello
$ hello
@botname hello
```

and it'll work.

@lins05 if this direction is something you think would be worthwhile in upstream, I can also add some unit tests here to complete this PR.

inspired by a previous stab at solving this problem by Dan

thanks!
-Dom

binary1230 committed:
add support for aliases
- you can use shorthand notation or aliases for talking to the bot now

lins05 commented on slackbot/dispatcher.py:
```diff
@@ -10,10 +10,14 @@
+never_matches_anything = 'a^'
+regex_aliases = '|'.join([re.escape(s) for s in settings.ALIASES.split(',')]) if hasattr(settings, 'ALIASES') else never_matches_anything
+
+AT_MESSAGE_MATCHER = re.compile(r'^(?:\<@(\w+)\>|({})):? (.*)$'.format(regex_aliases))
```
I'd rather do it like this:

``` python
if hasattr(settings, 'ALIASES'):
   AT_MESSAGE_MATCHER = ...
else:
  AT_MESSAGE_MATCHER = ...
```

The `never_matches_anything` trick is cool, but not so good for code readability.

lins05 commented:
@binary1230 Thanks for the patch! I like the idea of using alias. The change mostly lgtm, would you please add a test for it?

binary1230 commented:
Cool, re-reading this you're right about the readability.  we'll make the change and add a test.

CC @migetman9

Thanks!
-Dom

Commit:
Alias Test

Commit:
Test Driver Function

driver.py references a variable that doesn’t exist.

binary1230 committed:
refine support for aliases

per code review, eliminate the concept of 'never_matches_anything' and make code more readable

binary1230 committed:
Merge remote-tracking branch 'origin/alias_support' into alias_support

binary1230 commented:
@lins05 ok per code review, tried another approach which eliminates `never_matches_anything`.  because aliases might not be present, I had to switch the regex match code from using the index of the group match to instead use a named group, which is a python-specific regex extension.  I think the end result is more readable, let me know what you think I'll be happy to make additional changes.

@migetman9 can you re-run the unit tests, make sure it passes again?

Thanks!

binary1230 commented:
also I think @migetman9 found a small bug in an error message in the test code, which is fixed in here too.

I will try and get my own testing setup going later today.  We should make sure the test covers both when we are and are not using aliases, if possible.

binary1230 committed:
create more robust unit testing for dispatcher, fix bug in dispatcher

 - create unit tests and fixtures to fully cover new alias feature
 - enable testing of dispatcher without having to connect to a live slack instance

binary1230 commented on slackbot/dispatcher.py:
```diff
@@ -22,6 +21,12 @@ def __init__(self, slackclient, plugins):
+            alias_regex = '|(?P<alias>{})'.format('|'.join([re.escape(s) for s in settings.ALIASES.split(',')]))
+
+        self.AT_MESSAGE_MATCHER = re.compile(r'^(?:\<@(?P<atuser>\w+)\>{}):? (?P<text>.*)$'.format(alias_regex))
+
```
new stuff: needed to move this into the `__init__` method in order to allow unit tests to re-create AT_MESSAGE_MATCHER when we create and destroy MessageDispatcher() in py.text fixtures

lins05 replied:
+1

binary1230 commented on slackbot/dispatcher.py:
```diff
@@ -69,24 +74,37 @@ def _on_new_message(self, msg):
 
+    def _get_bot_id(self):
+        return self._client.login_data['self']['id']
+
```
needed to make this a separate method in order to allow monkeypatch to fake the bot ID in unit tests

binary1230 commented on slackbot/dispatcher.py:
```diff
@@ -69,24 +74,37 @@ def _on_new_message(self, msg):
-            m = AT_MESSAGE_MATCHER.match(text)
             if m:
-                msg['text'] = m.group(2)
+                msg['text'] = m.groupdict().get('text', None)
```
fixed legit bug here uncovered during unit tests

binary1230 commented on slackbot/settings.py:
```diff
@@ -11,6 +11,16 @@
+'''
+ALIASES = ''
+
+'''
```
it became easier for unit tests to have this be set to a default value of ''

binary1230 commented on tests/functional/driver.py:
```diff
@@ -247,7 +247,7 @@ def _join_test_channel(self):
         else:
             raise RuntimeError('Have you created the private group {} for testing?'.format(
-                self.group_name))
+                self.test_group))
```
legit bug found by @migetman9

binary1230 commented on tests/functional/test_functional.py:
```diff
@@ -156,3 +156,7 @@ def test_bot_reply_with_unicode_message(driver):
+
+def test_bot_reply_with_alias_message(driver):
+    driver.send_channel_message("! hello", False, False)
+    driver.wait_for_bot_channel_message("hello sender!", tosender=True)
```
this is useful to have here though @migetman9 we need to ensure something sets ALIASES either here in the code for the driver, or in whatever script runs the functional testing

lins05 replied:
You can set the aliases for the test by adding it to `tests/functional/settings.py`, which would be imported by the bot instance started by the test.

binary1230 commented on tests/unit/test_dispatcher.py:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    msg = dispatcher.filter_text(msg)
+    assert msg['text'] == 'hello'
```
went a little nuts here, but these tests ensure 100% code coverage for MessageDispatcher.filter_text(), including working with and without aliases.  I know that the functional testing covers some of this already, but I wanted to add these tests to make sure our alias code was 100% rock solid.  I ran with code coverage and filter_text() is fully covered in cases of both aliases and non-aliases.

lins05 commented on slackbot/dispatcher.py:
```diff
@@ -22,6 +21,12 @@ def __init__(self, slackclient, plugins):
         self._plugins = plugins
 
+        alias_regex = ''
+        if hasattr(settings, 'ALIASES') and settings.ALIASES != '':
```
This can be simplified to:

``` python
if getattr(settings, 'ALIASES', None):
...
```

lins05 commented on tests/unit/test_dispatcher.py:
```diff
@@ -0,0 +1,94 @@
+
+aliases = ['!', '$', 'botbro']
+fake_bot_id = 'US99999'
+fake_bot_name = '<@' + fake_bot_id + '>'
```
I'd prefer to use ALL CAPITAL letters for these module level constants.

binary1230 replied:
fixed

binary1230 committed:
simplify getting attribute

binary1230 committed:
capitalize constants in unit tests
- per code review

binary1230 commented:
fixed capitalization and attribute, @migetman9, can you look at the functional test stuff when you get a chance

Commit:
add ALIASES attribute for test_functional.py

daredoes commented:
test_functional should work now

binary1230 commented:
cool, I'm happy with all of this, if it looks good to you @lins05 then we're solid.
